compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -156,15 +156,23 @@ enum AdjustMode {
 /// this last case, so we need to throw an error ourselves.
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
 enum MutblCap {
-    /// Mutability restricted to immutable;
-    /// contained span, if present, should be shown in diagnostics as the reason.
-    Not(Option<Span>),
+    /// Mutability restricted to immutable.
+    ///
+    /// The contained span, if present, points to an `&` pattern
+    /// that is the reason for the restriction,
+    /// and which will be reported in a diagnostic.
+    /// (Said diagnostic is shown only if
+    /// replacing the `&` pattern with `&mut` would allow the code to compile.)
+    ///
+    /// (Outer [`Option`] is for whether to show the diagnostic,
+    /// inner [`Option`] is for whether we have a span we can report)
+    Not(Option<Option<Span>>),
```

Comment:
I am really not a fan of `Option<Option<`. Could you make the outer `Option` into a different variant of `MutblCap`? And split `cap_mutbl_to_not` into two functions to match